flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -182,7 +181,7 @@
 	protected final TimerService timerService;
 
 	/** The currently active background materialization threads. */
-	private final CloseableRegistry cancelables = new CloseableRegistry();
+	private final AsyncCheckpointRunnableRegistry cancelables = new AsyncCheckpointRunnableRegistry();
```

Comment:
This registry stores not only `AsyncCheckpointRunnableRegistry` but also `SubtaskCheckpointCoordinator` (and other objects).

Setting it to `AsyncCheckpointRunnableRegistry` makes me think it only stores Runnables and should go into `SubtaskCheckpointCoordinatorImpl`.

I'd suggest to explicitly keep track of async checkpointing tasks in `SubtaskCheckpointCoordinatorImpl`.

Reply:
In the beginning, I plan to wrap `CloseableRegistry` and found I have to use another lock to synchronized. That's why I currently have to extend `CloseableRegistry` to leverage the same `getSynchronizationLock()`.
And I agree the name of `AsyncCheckpointRunnableRegistry` is misleading, I'll prepare to wrap `CloseableRegistry` again as I think another internal lock should not harm much.

Reply:
But why not to store `AsyncCheckpointRunnable`s in `SubtaskCheckpointCoordinatorImpl`?
That is, move `Map<Long, AsyncCheckpointRunnable> checkpoints` from `AsyncCheckpointRunnableRegistry` to `SubtaskCheckpointCoordinatorImpl`.

That would make it more clear and flexible (we could change the type in map, we could avoid passing registry to `SubtaskCheckpointCoordinatorImpl` etc.).

Reply:
Because `asyncCheckpointRunnable` is registered to `closeableRegistry` when it begins to run asynchronously, we have to introduce something to wrap the `checkpointId`, `AsyncCheckpointRunnable` and `closeableRegistry` to let cancellation available, which is the role `AsyncCheckpointRunnableRegistry` played. Even we move `Map<Long, AsyncCheckpointRunnable> checkpoints` from `AsyncCheckpointRunnableRegistry` to `SubtaskCheckpointCoordinatorImpl`, we still need something to warp them all.

Moreover, since we register the `closeableRegistry` within `AsyncCheckpointRunnable`, and we create the runnable in `SubtaskCheckpointCoordinatorImpl`, I'm afraid it's not easy to avoid passing registry to `SubtaskCheckpointCoordinatorImpl`.

Reply:
> we still need something to warp them all.

Do you mean a scenario when all runnables need to be closed? 
For this we can explicitly iterate over a map and close them in `SubtaskCheckpointCoordinatorImpl.close()` (it's already registered with `closeableRegistry`).


> it's not easy to avoid passing registry

I didn't mean we should do it now but it's one thing we can consider in future. Having it tied to runnables makes it a bit harder.

Reply:
Actually, current `AsyncCheckpointRunnableRegistry` is a wrapper of `closeableRegistry` and ` Map<Long, AsyncCheckpointRunnable> checkpoints`.
And if we do not use a wrapper but with two separate `closeableRegistry` and `checkpoints`,  the `AsyncCheckpointRunnable` would register itself to `closeableRegistry` and put itself with checkpoint id to `checkpoints` when begin async phase. 
Once it reach the finally phase of the runnable, it will remove from the `checkpoints` and then unregister from `closeableRegistry`. Actually, I cannot see any real difference compared with current `AsyncCheckpointRunnableRegistry`.
Perhaps the name of `AsyncCheckpointRunnableRegistry` could easily mislead developers? And a name like wrapper is better?

Reply:
I think that `SubtaskCheckpointCoordinator` shouldn't delegate *its* responsibility of maintaining _its_ runnables to `StreamTask.closeableRegistry`. Changing the names only masquerades it.

Using the same registry creates unnecessary connection, making it more difficult to change `StreamTask` (for example, to remove registry type).

At runtime, I guess behavior is more or less the same.